src/anomalib/training/learning_strategies/default/test.py

```diff
@@ -0,0 +1,87 @@
+"""Test loop."""
+
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+from functools import lru_cache
+from typing import Any, List, Optional
+
+from pytorch_lightning.loops.dataloader.evaluation_loop import EvaluationLoop
+from pytorch_lightning.loops.epoch.evaluation_epoch_loop import EvaluationEpochLoop
+from pytorch_lightning.utilities.types import EPOCH_OUTPUT, STEP_OUTPUT
+
+import anomalib.training as core
+
+
+class AnomalibTestEpochLoop(EvaluationEpochLoop):
+    def __init__(self) -> None:
+        super().__init__()
+        self.trainer: core.AnomalibTrainer
+
+    def _evaluation_step(self, **kwargs: Any) -> Optional[STEP_OUTPUT]:
+        """Runs the ``test_step``."""
+        outputs = super()._evaluation_step(**kwargs)
+        if outputs is not None:
+            self.trainer.post_processor.outputs_to_cpu(outputs)
+            self.trainer.post_processor.compute_labels(outputs)
+            self.trainer.post_processor.apply_thresholding(self.trainer.lightning_module, outputs)
+            self.trainer.normalizer.normalize(self.trainer.lightning_module, outputs)
+        return outputs
+
+    def _evaluation_step_end(self, *args, **kwargs) -> STEP_OUTPUT | None:
+        """Runs ``test_step_end`` after the end of one test step."""
+        outputs = super()._evaluation_step_end(*args, **kwargs)
+        # Add your code here
+        return outputs
+
+    def _on_evaluation_batch_end(self, output: Optional[STEP_OUTPUT], **kwargs: Any) -> None:
+        """The ``on_test_batch_end`` hook.
+
+        Args:
+            output: The output of the performed step
+            batch: The input batch for the step
+            batch_idx: The index of the current batch
+            dataloader_idx: Index of the dataloader producing the current batch
+        """
+        super()._on_evaluation_batch_end(output, **kwargs)
+
+    @lru_cache(1)
+    def _should_track_batch_outputs_for_epoch_end(self) -> bool:
+        """Track batch outputs for epoch end.
+
+        If this is not overridden, the outputs are not collected if the model does not have a ``test_step_end``
+        method. This ensures that the outputs are collected even if the model does not have a ``test_step_end`` method.
+        """
+        return True
```

Comment:
Now that I think about it, I think we can remove this. Originally this was because the `test_step_end` is not called when the lightning module's `test_step` is not present. But if the `test_step` is not overriden then we don't get any outputs in `test_step_end` so it is pretty useless anyways. I'll remove this.

Reply:
I feel it would be better to do it in `test_step_end`. What do you mean with `the outputs are deleted`? It's not possible to test with a PL model without implementing the `test_step_end` method?